### PR TITLE
Bug 1761485 - Expose ErrorType through its own entry point.

### PR DESCRIPTION
**Pull request by tawahpeggy**
### Pull Request checklist ###

- [ ] **Quality**: Make sure this PR builds and runs cleanly.
  - Inside the `glean/` folder, run:
    - `npm run test` Runs _all_ tests
    - `npm run lint` Runs _all_ linters
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry to `CHANGELOG.md` or an explanation of why it does not need one
- [ ] **Documentation**: This PR includes documentation changes, an explanation of why it does not need that or a follow-up bug has been filed to do that work
  - Documentation should be added to [The Glean Book](https://mozilla.github.io/glean/book/index.html) when making changes to Glean's user facing API
    - When changes to the Glean Book are necessary, link to the corresponding PR on the [`mozilla/glean`](https://github.com/mozilla/glean) repository
  - Documentation should be added to [the Glean.js developer documentation](https://github.com/mozilla/glean.js/tree/main/docs) when making internal code changes

### CODE REVIEW

File: glean/package-lock.json [hunk 1]
```diff
@@ -1,7886 +1,24 @@
 {
   "name": "@mozilla/glean",
   "version": "1.0.0",
-  "lockfileVersion": 2,
+  "lockfileVersion": 1,
   "requires": true,
-  "packages": {
```
Comment (Dexterp37) [inline glean/package-lock.json:6]: Looks like you're using an outdated `npm` and the lockfile was rewritten with an older version. Would you kindly update your local npm and drop this change?

File: glean/package.json [hunk 2]
```diff
@@ -7,6 +7,7 @@
   "exports": {
     "./package.json": "./package.json",
     "./private/metrics/*": "./dist/core/metrics/types/*.js",
+    "./error": "./dist/core/error/error_type.js",
```
Comment (Dexterp37) [inline glean/package.json:10]: Would you kindly make this the first entry in this object? (i.e. move this before `.package.json`)

File: glean/tests/unit/platform/utils/webext/sample/package-lock.json [hunk 3]
```diff
@@ -1,7381 +1,8 @@
 {
   "name": "web-extension",
   "version": "1.0.0",
-  "lockfileVersion": 2,
+  "lockfileVersion": 1,
```
Comment (Dexterp37) [inline glean/tests/unit/platform/utils/webext/sample/package-lock.json:4]: Same consideration as the other lockfile. Would you kindly drop this change?

**Review by Dexterp37 — CHANGES_REQUESTED** [review]
Hi @tawahpeggy and thank you for your contribution!

This needs a few changes before we can merge it: because you're using an old version of npm, the lock files have been rewritten in an old format, making CI fail.

In addition to the above, would you kindly add an entry for this change at the top of the `CHANGELOG.md` file?

**Comment by tawahpeggy** [comment]
Thank you very much for your review, will update the PR ASAP

File: glean/package.json [hunk 4]
```diff
@@ -96,18 +96,18 @@
...
-    "@types/selenium-webdriver": "^4.0.14",
+    "@types/node": "^17.0.25",
+    "@types/selenium-webdriver": "^4.0.19",
     "@types/sinon": "^10.0.0",
     "@types/sqlite3": "^3.1.7",
     "@types/uuid": "^8.3.1",
-    "@typescript-eslint/eslint-plugin": "^5.0.0",
-    "@typescript-eslint/parser": "^5.0.0",
-    "eslint": "^8.0.1",
+    "@typescript-eslint/eslint-plugin": "^5.20.0",
```
Comment (Dexterp37) [inline glean/package.json:104]: @tawahpeggy while it's eventually desirable, we should not be updating the versions of the dependencies as part of this PR if possible. Thanks!

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,7 @@
 
 [Full changelog](https://github.com/mozilla/glean.js/compare/v1.0.0...main)
 
+* [#1318](https://github.com/mozilla/glean.js/pull/1318):Expose ErrorType through its own entry point.
```
Comment (Dexterp37) [inline CHANGELOG.md:5]:
```suggestion
* [#1318](https://github.com/mozilla/glean.js/pull/1318): Expose `ErrorType` through its own entry point.
```

**Review by Dexterp37 — CHANGES_REQUESTED** [review]

**Comment by tawahpeggy** [comment]
Hello @Dexterp37, please do you have an idea why some tests fail?

**Comment by Dexterp37** [comment]
> Hello @Dexterp37, please do you have an idea why some tests fail?

It doesn't seem to be able to import `ErrorType`. I believe the imports need to be adjusted :)

```
TSError: ⨯ Unable to compile TypeScript:
tests/unit/index.test.ts:6:10 - error TS2614: Module '"@mozilla/glean/webext"' has no exported member 'ErrorType'. Did you mean to use 'import ErrorType from "@mozilla/glean/webext"' instead?

6 import { ErrorType } from "@mozilla/glean/webext";
           ~~~~~~~~~

    at createTSError (/home/circleci/project/samples/browser/webext/node_modules/ts-node/src/index.ts:820:12)
    at reportTSError (/home/circleci/project/samples/browser/webext/node_modules/ts-node/src/index.ts:824:19)
    at getOutput (/home/circleci/project/samples/browser/webext/node_modules/ts-node/src/index.ts:1014:36)
    at Object.compile (/home/circleci/project/samples/browser/webext/node_modules/ts-node/src/index.ts:1322:43)
    at transformSource (/home/circleci/project/samples/browser/webext/node_modules/ts-node/src/esm.ts:380:37)
    at load (/home/circleci/project/samples/browser/webext/node_modules/ts-node/src/esm.ts:280:51)
    at async ESMLoader.load (node:internal/modules/esm/loader:359:20)
    at async ESMLoader.moduleProvider (node:internal/modules/esm/loader:280:47)
    at async link (node:internal/modules/esm/module_job:70:21)

Exited with code exit status 1

CircleCI received exit code 1

```

File: glean/package-lock.json (hunk 1 above)
Comment (Dexterp37) [reply glean/package-lock.json:6]: @tawahpeggy this was fully removed instead of restored, would you kindly restore the file?

**Comment by tawahpeggy** [comment]
> > Hello @Dexterp37, please do you have an idea why some tests fail?
> 
> It doesn't seem to be able to import `ErrorType`. I believe the imports need to be adjusted :)
> 
> ```
> TSError: ⨯ Unable to compile TypeScript:
> tests/unit/index.test.ts:6:10 - error TS2614: Module '"@mozilla/glean/webext"' has no exported member 'ErrorType'. Did you mean to use 'import ErrorType from "@mozilla/glean/webext"' instead?
> 
> 6 import { ErrorType } from "@mozilla/glean/webext";
>            ~~~~~~~~~
> 
>     at createTSError (/home/circleci/project/samples/browser/webext/node_modules/ts-node/src/index.ts:820:12)
>     at reportTSError (/home/circleci/project/samples/browser/webext/node_modules/ts-node/src/index.ts:824:19)
>     at getOutput (/home/circleci/project/samples/browser/webext/node_modules/ts-node/src/index.ts:1014:36)
>     at Object.compile (/home/circleci/project/samples/browser/webext/node_modules/ts-node/src/index.ts:1322:43)
>     at transformSource (/home/circleci/project/samples/browser/webext/node_modules/ts-node/src/esm.ts:380:37)
>     at load (/home/circleci/project/samples/browser/webext/node_modules/ts-node/src/esm.ts:280:51)
>     at async ESMLoader.load (node:internal/modules/esm/loader:359:20)
>     at async ESMLoader.moduleProvider (node:internal/modules/esm/loader:280:47)
>     at async link (node:internal/modules/esm/module_job:70:21)
> 
> Exited with code exit status 1
> 
> CircleCI received exit code 1
> ```

Hmmmmm this is strange and I do not even know what I have done wrong.
everything is underlined in the test file :sad: .
image

**Comment by Dexterp37** [comment]
> Hmmmmm this is strange and I do not even know what I have done wrong. everything is underlined in the test file :sad: . 

In this PR you're changing the way the `ErrorType` symbol is being exported. Without this PR, things are being exported via the platform-specific endpoints, e.g. `import { ErrorType } from "@mozilla/glean/webext";`.
This PR exposes it through its own entry point and so call sites need to be adjusted like this:

`import { ErrorType } from "@mozilla/glean/error";`

**Comment by Dexterp37** [comment]
@tawahpeggy note that the proper import need to be applied to all the samples and in all the other locations within the codebase.

File: glean/package.json [hunk 6]
```diff
@@ -4,7 +4,8 @@
   "description": "An implementation of the Glean SDK, a modern cross-platform telemetry client, for JavaScript environments.",
   "type": "module",
   "sideEffects": false,
-  "exports": {
+  "exports": {
```
Comment (Dexterp37) [inline glean/package.json:7]: nit: please remove the trailing whitespaces on this line

File: glean/src/core/error/index.ts [hunk 7]
```diff
@@ -3,7 +3,7 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 import type { MetricType } from "../metrics/index.js";
-import type { ErrorType } from "./error_type.js";
```
Comment (Dexterp37) [inline glean/src/core/error/index.ts:6]: I'm not sure this change is required: this is internal, and so we can directly reference the file as we did before. We only need to change occurrences of imports that import the "exposed" entrypoint.

**Review by Dexterp37 — CHANGES_REQUESTED** [review]

File: glean/src/core/error/index.ts [hunk 8]
```diff
@@ -3,7 +3,7 @@
...
+import { ErrorType } from "./error_type.js";
```
Comment (Dexterp37) [inline glean/src/core/error/index.ts:6]: Would you kindly add back the `type`?

File: glean/src/entry/qt.ts [hunk 9]
```diff
@@ -2,7 +2,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-import { ErrorType } from "../core/error/error_type.js";
+import type { ErrorType } from "@mozilla/glean/error";
```
Comment (Dexterp37) [inline glean/src/entry/qt.ts:5]: Was this change required?

**Review by Dexterp37 — CHANGES_REQUESTED** [review]
We're almost there! I think with the last two requested changes it should be ok for merging.

**Comment by tawahpeggy** [comment]
Hello @Dexterp37, hope you are having a great day.
please help check this PR and see if there's something else to be done

**Comment by Dexterp37** [comment]
> Hello @Dexterp37, hope you are having a great day. please help check this PR and see if there's something else to be done

Hi @tawahpeggy , I'm doing great, thanks. Note that the PR is still failing the tests (see the failing CI jobs), so we can't merge it: would you kindly fix the codebase to make sure tests run ok? You can also run them locally to reduce the turnaround time.

**Comment by tawahpeggy** [comment]
> > Hello @Dexterp37, hope you are having a great day. please help check this PR and see if there's something else to be done
> 
> Hi @tawahpeggy , I'm doing great, thanks. Note that the PR is still failing the tests (see the failing CI jobs), so we can't merge it: would you kindly fix the codebase to make sure tests run ok? You can also run them locally to reduce the turnaround time.

I noticed the failed test but I really do not know how to fix it. any heads up?

**Comment by Dexterp37** [comment]
> > > Hello @Dexterp37, hope you are having a great day. please help check this PR and see if there's something else to be done
> > 
> > 
> > Hi @tawahpeggy , I'm doing great, thanks. Note that the PR is still failing the tests (see the failing CI jobs), so we can't merge it: would you kindly fix the codebase to make sure tests run ok? You can also run them locally to reduce the turnaround time.
> 
> I noticed the failed test but I really do not know how to fix it. any heads up?

I think we need to add something like this

```
      "error": [
        "./dist/types/core/error/error_type.d.ts"
      ]
```

after `"testing"` in [here](https://github.com/mozilla/glean.js/blob/e7e5bd3a19f81082f9c0cba08d6e0ad4f817c67f/glean/package.json#L43)

**Comment by tawahpeggy** [comment]
Thank you @Dexterp37, 
I get this message at the end. when I run  the test, please is it something I should worry about? image

**Comment by moz-glean** [comment]
# Build size report

Merging https://github.com/mozilla/glean.js/pull/1318 into [main](https://github.com/mozilla/glean.js) will:
* **Leave** the size of full **Web Extension** bundle **unchanged**.
* **Leave** the size of full **Website** bundle **unchanged**.
* **Leave** the size of full **Node.js** bundle **unchanged**.
* **Leave** the size of full **Qt/QML** bundle **unchanged**.

---

| | Current size | New size | Size increase |
|--:|:---:|:---:|:---:|
| **Web Extension** |
| core only | 55 KB | 55 KB | 📉 0 bytes |
| full bundle | 95 KB | 95 KB | 📉 0 bytes |
| **Website** |
| core only | 56 KB | 56 KB | 📉 0 bytes |
| full bundle | 96 KB | 96 KB | 📉 0 bytes |
| **Node.js** |
| core only | 55 KB | 55 KB | 📉 23 bytes |
| full bundle | 101 KB | 101 KB | 📉 23 bytes |
| **Qt/QML** |
| core only | 78 KB | 78 KB | 📉 0 bytes |
| full bundle | 78 KB | 78 KB | 📉 0 bytes |

**Review by Dexterp37 — APPROVED** [review]

**Comment by tawahpeggy** [comment]
> > Thank you @Dexterp37, I get this message at the end. when I run the test, please is it something I should worry about? image
> 
> I don't think we should worry about that as long as this runs on CI. Let me check :)

Okay understood thanks.

**Pull request merged**